iwyu_ast_util.h

```diff
@@ -697,6 +697,13 @@ const clang::Type* Desugar(const clang::Type* type);
 // vector<pair<Foo, Bar>> has components pair<Foo,Bar>, Foo, and Bar.
 set<const clang::Type*> GetComponentsOfType(const clang::Type* type);
 
+// Returns types for determination of their "provision" status. They are
+// canonicalized because intermediate sugar should be always provided already
+// according to language rules. Substituted template parameter types (and their
+// underlying types) cannot be provided.
+std::set<const clang::Type*> GetComponentsOfTypeWithoutSubstituted(
```

Comment:
Could you add an example to contrast with the ones on lines 696-697? I find the "without substituted" naming a little opaque, but mostly because I don't understand all the AST details around template instantiation. So for the `Foo*` and `vector<Foo>` cases, will we still get `Foo`? And for `vector<pair<Foo,Bar>>` will we get the same as above (assuming `pair` is not a typedef? Might be interesting to expand that example with type aliases to show the difference.

Reply:
I've added a couple of examples. It's worth noting that these functions return not only "components" but also the given original type. It is exactly what is needed for the purpose of this PR. But, maybe, naming of the functions could be improved somehow...